R/data_summary.R

```diff
@@ -0,0 +1,217 @@
+#' grouping variable.
+#' @param ... One or more named expressions that define the new variable name
+#' and the function to compute the summary statistic. Example:
+#' `mean_sepal_width = mean(Sepal.Width)`. The expression can also be provided
+#' as a character string, e.g. `"mean_sepal_width = mean(Sepal.Width)"`. The
+#' summary function `n()` can be used to count the number of observations.
+#'
+#' @seealso `value_at()`, which can be used inside a `data_summary()` call.
+#'
+#' @return A data frame with the requested summary statistics.
+#'
+#' @examples
+#' data(iris)
+#' data_summary(iris, MW = mean(Sepal.Width), SD = sd(Sepal.Width))
+#' data_summary(
+#'   iris,
+#'   MW = mean(Sepal.Width),
+#'   SD = sd(Sepal.Width),
+#'   by = "Species"
+#' )
+#'
+#' # same as
+#' d <- data_group(iris, "Species")
+#' data_summary(d, MW = mean(Sepal.Width), SD = sd(Sepal.Width))
+#'
+#' # multiple groups
+#' data(mtcars)
+#' data_summary(mtcars, MW = mean(mpg), SD = sd(mpg), by = c("am", "gear"))
+#'
+#' # expressions can also be supplied as character strings
+#' data_summary(mtcars, "MW = mean(mpg)", "SD = sd(mpg)", by = c("am", "gear"))
+#'
+#' # count observations within groups
+#' data_summary(mtcars, observations = n(), by = c("am", "gear"))
+#'
+#' # first and last observations of "mpg" within groups
+#' data_summary(
+#'   mtcars,
+#'   first = value_at(mpg),
+#'   last = value_at(mpg, -1),
+#'   by = c("am", "gear")
+#' )
+#' @export
+data_summary <- function(x, ...) {
+  UseMethod("data_summary")
+}
+
+
+#' @export
+data_summary.matrix <- function(x, ..., by = NULL, include_na = TRUE) {
+  data_summary(as.data.frame(x), ..., by = by, include_na = include_na)
+}
+
+
+#' @export
+data_summary.default <- function(x, ...) {
+  insight::format_error("`data_summary()` only works for (grouped) data frames and matrices.")
+}
+
+
+#' @rdname data_summary
+#' @export
+data_summary.data.frame <- function(x, ..., by = NULL, include_na = TRUE) {
+  dots <- eval(substitute(alist(...)))
+
+  # do we have any expression at all?
+  if (length(dots) == 0) {
+    insight::format_error("No expressions for calculating summary statistics provided.")
+  }
+
+  if (is.null(by)) {
+    # when we have no grouping, just compute a one-row summary
+    summarise <- .process_datasummary_dots(dots, x)
+    out <- data.frame(summarise)
+    colnames(out) <- vapply(summarise, names, character(1))
+  } else {
+    # sanity check - is "by" a character string?
+    if (!is.character(by)) {
+      insight::format_error("Argument `by` must be a character string, indicating the name of a variable in the data.")
+    }
+    # is "by" in the data?
+    if (!all(by %in% colnames(x))) {
+      by_not_found <- by[!by %in% colnames(x)]
+      insight::format_error(
+        paste0("Variable \"", by_not_found, "\" not found in the data."),
+        .misspelled_string(colnames(x), by_not_found, "Possibly misspelled?")
+      )
```

Comment:
Currently this prints the same message for each unknown variable:
```r
> data_summary(mtcars, MW = mean(mpg), SD = sd(mpg), by = c("bam", "geas"))
Error: Variable "bam" not found in the data.
Variable "geas" not found in the data.
Did you mean one of "am" or "gear"?
```

Can you aggregate those messages to have `Variable 'bam' and 'geas' not found in the data` instead?